### ISSUE TITLE
JSD Loss

### ISSUE THREAD

**Issue by minsub0922** [minsub0922/gpt_portal#5]
```
from typing import Optional, Dict, Any, Tuple
import torch
import torch.nn.functional as F
from transformers import Trainer


def _get_logits(outputs) -> torch.Tensor:
    """
    Hugging Face 모델 출력에서 logits만 뽑아 [B, L, V] 형태로 맞춘다.
    - 분류: [B, V] -> [B, 1, V]
    - LM:   [B, L, V] 그대로
    """
    logits = outputs["logits"] if isinstance(outputs, dict) else outputs.logits
    if logits.dim() == 2:
        # [B, V] -> [B, 1, V]
        logits = logits.unsqueeze(1)
    return logits


def _get_attention_mask(inputs: Dict[str, torch.Tensor], logits: torch.Tensor) -> torch.Tensor:
    """
    [B, L] attention_mask가 없으면 1로 채움.
    logits에 맞춰 L 길이로 broadcast.
    """
    if "attention_mask" in inputs and inputs["attention_mask"] is not None:
        attn = inputs["attention_mask"]
        if attn.dim() == 1:
            attn = attn.unsqueeze(0)
    else:
        B, L, _ = logits.shape
        attn = torch.ones(B, L, device=logits.device, dtype=torch.long)
    return attn


def _masked_mean(x: torch.Tensor, mask: Optional[torch.Tensor], eps: float = 1e-12) -> torch.Tensor:
    """
    x: [B, L] 또는 스칼라. mask: [B, L].
    """
    if mask is None:
        return x.mean()
    while mask.dim() < x.dim():
        mask = mask.unsqueeze(-1)
    s = (x * mask).sum()
    c = mask.sum().clamp_min(eps)
    return s / c


def jsd_loss_from_logits(
    student_logits: torch.Tensor,   # [B, L, V]
    teacher_logits: torch.Tensor,   # [B, L, V]
    attn_mask: Optional[torch.Tensor] = None,  # [B, L]
    beta: float = 0.5,
    T_s: float = 1.0,
    T_t: float = 1.0,
    eps: float = 1e-8,
    topk: Optional[int] = None,
    topk_from: str = "teacher",     # {"teacher","student"}, subset softmax 근사
) -> torch.Tensor:
    """
    JSD_beta = beta * KL(P_T || M) + (1-beta) * KL(P_S || M),
      where M = beta * P_T + (1-beta) * P_S

    메모리/속도 최적화를 위해 top-k 근사(선택)를 제공:
      - topk가 주어지면 지정된 분포의 상위 k 로짓으로 softmax 후
        re-normalize 하는 'subset softmax' 근사(실전 QAT에서 자주 쓰는 테크닉).
      - 정확도가 최우선이면 topk=None 로 전체 vocabulary에서 계산.
    """
    assert student_logits.shape == teacher_logits.shape, "student/teacher logits shape mismatch"
    B, L, V = student_logits.shape

    if topk is not None and 0 < topk < V:
        if topk_from not in {"teacher", "student"}:
            raise ValueError("topk_from must be 'teacher' or 'student'")
        ref = teacher_logits if topk_from == "teacher" else student_logits
        idx = torch.topk(ref, k=topk, dim=-1).indices  # [B, L, k]

        s = torch.gather(student_logits, -1, idx)
        t = torch.gather(teacher_logits, -1, idx)

        P_s = F.softmax(s / T_s, dim=-1).clamp_min(eps)
        P_t = F.softmax(t / T_t, dim=-1).clamp_min(eps)
    else:
        P_s = F.softmax(student_logits / T_s, dim=-1).clamp_min(eps)
        P_t = F.softmax(teacher_logits / T_t, dim=-1).clamp_min(eps)

    M = (beta * P_t + (1.0 - beta) * P_s).clamp_min(eps)

    # KL(P || M) = sum P * (log P - log M)
    KL_t = (P_t * (P_t.log() - M.log())).sum(dim=-1)  # [B, L]
    KL_s = (P_s * (P_s.log() - M.log())).sum(dim=-1)  # [B, L]
    jsd = beta * KL_t + (1.0 - beta) * KL_s          # [B, L]

    return _masked_mean(jsd, attn_mask)


class JSDTrainer(Trainer):
    """
    KDTrainer와 동일 인터페이스를 유지하면서,
    teacher–student JSD 손실을 사용해 Distill-QAT를 수행하는 Trainer.

    추가 인자(모두 선택, 기본값 제공):
      - jsd_beta: JSD의 beta (기본 0.5; 대칭)
      - jsd_scale: 최종 손실에서 JSD 가중치
      - ce_scale:  라벨이 있을 때 CE 섞기 위한 가중치(기본 0.0 → JSD 단독)
      - temperature_student / temperature_teacher: softmax 온도
      - topk: 어휘가 큰 경우 subset softmax 근사용 k (None이면 정확 계산)
      - topk_from: "teacher" 또는 "student" (subset 기준 분포)
      - eps: 수치 안정성용 epsilon
      - detach_teacher: True면 teacher 그래디언트 완전 차단(기본 True)

    기존 KDTrainer 시그니처 호환:
      - teacher_model (필수)
      - l2l_loss_scale (유지용 인자; 기본 0.0, 현재 로짓 L2는 사용하지 않음)
    """

    def __init__(
        self,
        teacher_model,
        l2l_loss_scale: float = 0.0,
        *args,
        jsd_beta: float = 0.5,
        jsd_scale: float = 1.0,
        ce_scale: float = 0.0,
        temperature_student: float = 1.0,
        temperature_teacher: float = 1.0,
        topk: Optional[int] = None,
        topk_from: str = "teacher",
        eps: float = 1e-8,
        detach_teacher: bool = True,
        **kwargs,
    ):
        super().__init__(*args, **kwargs)
        self.teacher_model = teacher_model
        self.l2l_loss_scale = float(l2l_loss_scale)

        # Freeze teacher
        self.teacher_model.eval()
        for p in self.teacher_model.parameters():
            p.requires_grad_(False)

        # JSD configs
        self.jsd_beta = float(jsd_beta)
        self.jsd_scale = float(jsd_scale)
        self.ce_scale = float(ce_scale)
        self.T_s = float(temperature_student)
        self.T_t = float(temperature_teacher)
        self.topk = topk
        self.topk_from = topk_from
        self.eps = float(eps)
        self.detach_teacher = bool(detach_teacher)

    def _compute_student_outputs(self, model, inputs) -> Any:
        return model(**inputs)

    @torch.no_grad()
    def _compute_teacher_outputs(self, inputs) -> Any:
        # labels 등 teacher에 불필요한 키는 제거
        t_inputs = {k: v for k, v in inputs.items() if k != "labels"}
        return self.teacher_model(**t_inputs)

    def _cross_entropy_loss(
        self,
        student_logits: torch.Tensor,  # [B, L, V] or [B, 1, V]
        labels: torch.Tensor,          # [B, L] or [B]
        attn_mask: Optional[torch.Tensor] = None,
        ignore_index: int = -100,
    ) -> torch.Tensor:
        """
        HF 기본 규칙을 따르는 CE:
        - LM: [B, L, V] vs [B, L]
        - Cls: [B, 1, V] vs [B]
        - label == ignore_index는 무시
        """
        B, L, V = student_logits.shape
        if labels.dim() == 1:
            labels = labels.unsqueeze(1)
        # [B*L, V], [B*L]
        logits_flat = student_logits.reshape(B * L, V)
        labels_flat = labels.reshape(B * L)

        loss = F.cross_entropy(logits_flat, labels_flat, ignore_index=ignore_index, reduction="none")
        loss = loss.view(B, L)

        if attn_mask is not None:
            loss = _masked_mean(loss, attn_mask)
        else:
            loss = loss.mean()
        return loss

    def compute_loss(
        self,
        model,
        inputs,
        return_outputs: bool = False,
        num_items_in_batch: Optional[int] = None,  # HF Trainer 시그니처 유지
    ) -> Tuple[torch.Tensor, Any]:
        """
        total_loss = jsd_scale * JSD + ce_scale * CE
        """
        # Ensure teacher on same device
        if next(self.teacher_model.parameters()).device != next(model.parameters()).device:
            self.teacher_model.to(next(model.parameters()).device)

        # 1) Student forward
        student_outputs = self._compute_student_outputs(model, inputs)
        s_logits = _get_logits(student_outputs)  # [B, L, V]

        # 2) Teacher forward (no grad)
        if self.detach_teacher:
            with torch.no_grad():
                teacher_outputs = self._compute_teacher_outputs(inputs)
        else:
            teacher_outputs = self._compute_teacher_outputs(inputs)
        t_logits = _get_logits(teacher_outputs).detach()  # 안전하게 detach

        # 3) Mask & labels
        attn_mask = _get_attention_mask(inputs, s_logits)
        labels = inputs.get("labels", None)

        # 4) JSD
        jsd = jsd_loss_from_logits(
            s_logits, t_logits,
            attn_mask=attn_mask,
            beta=self.jsd_beta,
            T_s=self.T_s, T_t=self.T_t,
            eps=self.eps,
            topk=self.topk,
            topk_from=self.topk_from,
        )

        total = self.jsd_scale * jsd

        # 5) (선택) CE 추가
        if self.ce_scale and labels is not None:
            ce = self._cross_entropy_loss(s_logits, labels, attn_mask=attn_mask, ignore_index=-100)
            total = total + self.ce_scale * ce

        if return_outputs:
            return total, student_outputs
        return total
```
